ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/util/PendingPoolTest.java

```diff
@@ -159,6 +160,86 @@ public void add_finalizedBlock() {
     assertThat(requiredRootDroppedEvents).isEmpty();
   }
 
+  @Test
+  public void add_moreThanMaxItems() {
+    for (int i = 0; i < maxItems * 2; i++) {
+      final SignedBeaconBlock block =
+          dataStructureUtil.randomSignedBeaconBlock(currentSlot.longValue());
+      pendingPool.add(block);
+
+      final int expectedSize = Math.min(maxItems, i + 1);
+      assertThat(pendingPool.contains(block)).isTrue();
+      assertThat(pendingPool.size()).isEqualTo(expectedSize);
+    }
+
+    // Final sanity check
+    assertThat(pendingPool.size()).isEqualTo(15);
+  }
+
+  @Test
+  public void add_dontDropItemsWhenNewItemsIgnored() {
+    final List<SignedBeaconBlock> blocks = new ArrayList<>();
+    for (int i = 0; i < maxItems; i++) {
+      final SignedBeaconBlock block =
+          dataStructureUtil.randomSignedBeaconBlock(currentSlot.longValue());
+      pendingPool.add(block);
+      blocks.add(block);
+
+      final int expectedSize = Math.min(maxItems, i + 1);
+      assertThat(pendingPool.contains(block)).isTrue();
+      assertThat(pendingPool.size()).isEqualTo(expectedSize);
+    }
+    assertThat(pendingPool.size()).isEqualTo(maxItems);
+
+    // Add some blocks that should be ignored
+    SignedBeaconBlock ignoredBlock =
+        dataStructureUtil.randomSignedBeaconBlock(currentSlot.plus(futureTolerance).plus(1));
+    pendingPool.add(ignoredBlock);
+    ignoredBlock =
+        dataStructureUtil.randomSignedBeaconBlock(currentSlot.minus(historicalTolerance).minus(1));
+    pendingPool.add(ignoredBlock);
+
+    // All initial blocks should still be present
+    assertThat(pendingPool.size()).isEqualTo(maxItems);
+    blocks.forEach(b -> assertThat(pendingPool.contains(b)).isTrue());
+  }
+
+  @Test
+  public void add_shouldDropOldestItemsWhenEnforcingLimits() {
+    final UInt64 olderSlot = currentSlot;
+    final UInt64 newerSlot = currentSlot.plus(1);
+    SignedBeaconBlock block;
+
+    // Add older block
+    SignedBeaconBlock olderBlock = dataStructureUtil.randomSignedBeaconBlock(olderSlot);
+    pendingPool.add(olderBlock);
+    assertThat(pendingPool.contains(olderBlock)).isTrue();
+
+    // Add max newer blocks
+    for (int i = 0; i < maxItems; i++) {
+      block = dataStructureUtil.randomSignedBeaconBlock(newerSlot);
+      pendingPool.add(block);
+
+      assertThat(pendingPool.contains(block)).isTrue();
+      assertThat(pendingPool.size()).isLessThanOrEqualTo(maxItems);
+    }
+
+    // Older block should've been dropped
+    assertThat(pendingPool.contains(olderBlock)).isFalse();
+
+    // Add older block
+    olderBlock = dataStructureUtil.randomSignedBeaconBlock(olderSlot);
+    pendingPool.add(olderBlock);
+    assertThat(pendingPool.contains(olderBlock)).isTrue();
+    // Add newer block
+    block = dataStructureUtil.randomSignedBeaconBlock(newerSlot);
+    pendingPool.add(block);
+
+    // Older block should be dropped when newer block is added
```

Comment:
Should we be adding the older block in the first place? if it's older than the rest of the items in PendingItems?

Reply:
Not sure how much it matters ... but I could pretty easily switch it to ignore newly added items if they're older than the oldest item.  My reasoning was that this way we're always able to add new items to the pending pool.  

Thinking about possible attacks:
* With the current version (always add new items): 
  * You could flood the node with new messages and knock everything else out of the pending pool 
  * But you'd have to keep flooding the node with new messages to keep knocking out valid items
* With the version where we only keep the newest items:
  * You could potentially flood the node with future items and knock everything else out
  * Then for some period of time you could potentially block any other items from being added (until the slot ticks forward some number of times and you start to receive more messages in this range) without continuing to flood the node

It feels like the current option is a bit safer.